### PR TITLE
Add graphql authorization to API (66)

### LINKED ISSUE #66: Backend API has authorization flow
**Story:**
As a user, I want to ensure that data is protected.

**Acceptance Criteria:**
* JWT access tokens are processed on requests
* A user (`ClaimsPrincipal`) is loaded upon successful JWT validation
* The `RequestContext` has a `User` field
* GraphQL queries use `me` instead of specifying a user id:
```
{
  me {
    id
    firstName
  }
}
```

**Notes:**

* The .NET Core 3.0 previews include a project template for SPA authentication. Although the project is currently using 2.2, this might be a useful reference. https://docs.microsoft.com/en-us/aspnet/core/security/authentication/identity-api-authorization?view=aspnetcore-3.0

### PR DESCRIPTION
**Pull request by cailyncodes**
Integrates the GraphQL.Authorization library into Hedwig.

Creates `AuthenticatedUserRequirement` and `IsAuthenticatedPolicy` but does not apply them to any endpoints (because no frontend code exists just yet to forward the bearer token).

Adds `me` GraphQL query.

Should resolve #66.

### CODE REVIEW

**Comment by cailyncodes** [comment]
Should resolve #66

File: src/Hedwig/ClientApp/src/contexts/Login/LoginHOC.tsx [hunk 1]
```diff
@@ -176,6 +176,7 @@ const LoginProvider: React.FC<LoginProviderPropsType> = (
           // unset code, so we can do refresh token exchanges subsequently
           code = undefined;
           if (isFirstRequest) {
+            console.log(resp);
```
Comment (jlhogan) [inline src/Hedwig/ClientApp/src/contexts/Login/LoginHOC.tsx]: delete debug!

File: src/Hedwig/ServiceExtensions.cs [hunk 2]
```diff
@@ -79,5 +85,23 @@ public static void ConfigureGraphQL(this IServiceCollection services)
 				.AddDataLoader()
 				.AddUserContextBuilder<RequestContext>(RequestContext.RequestContextCreator);
 		}
+
+		public static void ConfigureAuthentication(this IServiceCollection services)
+		{
+			JwtSecurityTokenHandler.DefaultInboundClaimTypeMap.Clear();
+			services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
```
Comment (jlhogan) [inline src/Hedwig/ServiceExtensions.cs]: should this go inside an `if dev` just to make it clear we need to do work on configuring this for prod (when we know what prod will look like)

File: src/Hedwig/Startup.cs [hunk 3]
```diff
@@ -35,6 +39,7 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env)
             {
                 app.UseDeveloperExceptionPage();
                 app.UseCors("AllowAll");
+                IdentityModelEventSource.ShowPII = true;
```
Comment (jlhogan) [inline src/Hedwig/Startup.cs]: what does this do? comment?

File: src/Hedwig/ServiceExtensions.cs [hunk 4]
```diff
@@ -79,5 +85,23 @@ public static void ConfigureGraphQL(this IServiceCollection services)
...
+						options.Authority = "https://openid:5050";
+						options.Audience = "hedwig_backend";
+						options.BackchannelHttpHandler = new HttpClientHandler
+						{
+							ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator
+						};
+					});
+			services.AddSingleton<IAuthorizationEvaluator, AuthorizationEvaluator>();
+			services.AddTransient<IValidationRule, AuthorizationValidationRule>();
+			services.AddSingleton(s => Permissions.GetAuthorizationSettings());
```
Comment (jlhogan) [inline src/Hedwig/ServiceExtensions.cs]: this applies the authsettings, which includes the authed user requirement, as middleware to all requests?

File: src/Hedwig/Schema/RequestContext.cs [hunk 5]
```diff
@@ -16,22 +18,30 @@ namespace Hedwig.Schema
...
-    public class RequestContext
-    {      
+    public class RequestContext : IProvideClaimsPrincipal
+    {
         /// <summary>
         /// The creator function to pass to GraphQLBuilder.AddUserContextBuilder()
         /// </summary>
         /// <returns>an instantiated RequestContext</returns>
-        public static Func<HttpContext, RequestContext> RequestContextCreator = httpCtx => new RequestContext();
+        public static readonly Func<HttpContext, RequestContext> RequestContextCreator = httpCtx =>
```
Comment (jlhogan) [inline src/Hedwig/Schema/RequestContext.cs]: update Schema tests to include:
``` 
var httpCtx = new Mock<HttpContext>();
var userContext= new RequestContext(httpCtx.Object.User);
``` 
to reflect new constructor

File: src/Hedwig/Schema/Queries/UserQuery.cs [hunk 6]
```diff
@@ -19,6 +22,17 @@ public UserQuery(IUserRepository repository)
 					return repository.GetUserByIdAsync(id);
 				}
 			);
+			Field<UserType>(
+				"me",
+				arguments: new QueryArguments(),
```
[the remark this answers is not in the document]
Comment (cailyncodes) [reply src/Hedwig/Schema/Queries/UserQuery.cs]: There shouldn't be any arguments included on this query. Is there a different way I should denote that?

File: src/Hedwig/ClientApp/src/contexts/Login/LoginHOC.tsx (hunk 1 above)
Comment (cailyncodes) [reply src/Hedwig/ClientApp/src/contexts/Login/LoginHOC.tsx]: Oops -- ty!

File: src/Hedwig/Startup.cs (hunk 3 above)
Comment (cailyncodes) [reply src/Hedwig/Startup.cs]: It prints the URLs of endpoints and values of JWT claims when there is a mismatch in validation. Super helpful for debugging. Will add note.

File: src/Hedwig/ServiceExtensions.cs (hunk 4 above)
Comment (cailyncodes) [reply src/Hedwig/ServiceExtensions.cs]: Not quite -- this DIs them into the app which then we can later apply the policies to specific endpoints. I will do that once my PR I am currently working on lands -- currently the frontend doesn't include the access token in requests so if we protected the endpoints no data would ever load.

**Pull request merged**